### PR TITLE
CEGUI

### PR DESCRIPTION
**Pull request by AnastaZIuk**
Status:
- brdf example is **ready** to launch CEGUI (without meshes, pipelines stuff, etc - it will be done by Criss)
- **working on** adjusting CEGUI iterface classes to our API

### CODE REVIEW

**Comment by devshgraphicsprogramming** [comment]
My dear summer child, please call ex 07 something more appropriate

**Comment by devshgraphicsprogramming** [comment]
what the fuck are you doing deleting example 13?

File: examples_tests/CMakeLists.txt [hunk 1]
```diff
@@ -18,15 +18,14 @@ add_subdirectory(03.GPU_Mesh EXCLUDE_FROM_ALL)
...
+add_subdirectory(07.BRDF_Explorer EXCLUDE_FROM_ALL)
+endif()
 add_subdirectory(08.HardwareInstancing EXCLUDE_FROM_ALL)
 add_subdirectory(09.ColorSpaceTest EXCLUDE_FROM_ALL)
 add_subdirectory(10.AllocPerfTest EXCLUDE_FROM_ALL)
 add_subdirectory(11.RayCastCollision EXCLUDE_FROM_ALL)
 add_subdirectory(12.Splines EXCLUDE_FROM_ALL)
-if (IRR_BUILD_CEGUI)
-add_subdirectory(13.BRDF_Explorer EXCLUDE_FROM_ALL)
-endif()
```
Comment (devshgraphicsprogramming) [inline examples_tests/CMakeLists.txt:29]: NO, this is not what I asked for AT ALL

Example 13 stays as the BRDF Explorer

Example 7 is supposed to be a pure CEGUI-only example to show off GUI, nothing else

File: ext/CEGUI/CMakeLists.txt [hunk 2]
```diff
@@ -1,19 +1,33 @@
...
   ExtCEGUI.cpp
 	CEGUIOpenGLState.cpp
 	CEGUIOpenGLClip.cpp
 	Helpers.cpp
+	"IrrlichtBaW/src/ExtGeometryBuffer.cpp"
+	"IrrlichtBaW/src/ExtImageCodec.cpp"
+	"IrrlichtBaW/src/ExtRenderer.cpp"
+	"IrrlichtBaW/src/ExtTexture.cpp"
+	"IrrlichtBaW/src/ExtTextureTarget.cpp"
+	"IrrlichtBaW/src/ExtViewportTarget.cpp"
```
Comment (devshgraphicsprogramming) [inline ext/CEGUI/CMakeLists.txt:30]: no need for the `IrrlichtBaW` in the path

also don't add `include` and `src` to paths in `ext` just keep it all in one folder

File: ext/CEGUI/IrrlichtBaW/include/ExtGeometryBuffer.h [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+ *   included in all copies or substantial portions of the Software.
+ *
+ *   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ *   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ *   IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ *   OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ *   ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *   OTHER DEALINGS IN THE SOFTWARE.
+ ***************************************************************************/
```
Comment (devshgraphicsprogramming) [inline ext/CEGUI/IrrlichtBaW/include/ExtGeometryBuffer.h:26]: this needs to change slightly to reflect that we've added to this code under Apache 2.0, obvs keep the CEGUI notice

File: ext/CEGUI/IrrlichtBaW/include/ExtGeometryBuffer.h [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+#ifndef _CEGUIIrrlichtBawGeometryBuffer_h_
+#define _CEGUIIrrlichtBawGeometryBuffer_h_
```
Comment (devshgraphicsprogramming) [inline ext/CEGUI/IrrlichtBaW/include/ExtGeometryBuffer.h:28]: for extensions we use the following guard
`_IRR_EXT_EXTENSIONNAME_FILENAME_H_INCLUDED_`

File: ext/CEGUI/IrrlichtBaW/include/ExtImageCodec.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    }
+}
+
+namespace CEGUI
+{
+    //! ImageCodec object that loads data via image loading facilities in IrrlichtBaW.
+    class IrrlichtBaWImageCodec : public ImageCodec
+    {
+    public:
+        IrrlichtBaWImageCodec(irr::video::IVideoDriver& driver);
```
Comment (devshgraphicsprogramming) [inline ext/CEGUI/IrrlichtBaW/include/ExtImageCodec.h:47]: no, by pointer not reference

File: ext/CEGUI/IrrlichtBaW/include/ExtImageCodec.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+        // implement required function from ImageCodec.
+        Texture* load(const RawDataContainer& data, Texture* result);
+
+    protected:
+        irr::video::IVideoDriver& d_driver;
```
Comment (devshgraphicsprogramming) [inline ext/CEGUI/IrrlichtBaW/include/ExtImageCodec.h:53]: no, by pointer not reference

File: ext/CEGUI/IrrlichtBaW/include/ExtImageCodec.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace irr
+{
+    namespace video
+    {
+        class IVideoDriver;
+    }
+}
+
+namespace CEGUI
```
Comment (devshgraphicsprogramming) [inline ext/CEGUI/IrrlichtBaW/include/ExtImageCodec.h:41]: I think namespace on this should be `irr::ext::CEGUI`

File: ext/CEGUI/IrrlichtBaW/include/ExtRenderer.h [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+ ***************************************************************************/
+#ifndef _IRRLICHT_BAW_CEGUI_RENDERER_
+#define _IRRLICHT_BAW_CEGUI_RENDERER_
+
+#include <CEGUI/Renderer.h>
+#include <CEGUI/Size.h>
+#include <CEGUI/Vector.h>
+#include <CEGUI/Rect.h>
+
+#include <Irrlicht.h>
```
Comment (devshgraphicsprogramming) [inline ext/CEGUI/IrrlichtBaW/include/ExtRenderer.h:35]: its `"irrlicht.h"

File: ext/CEGUI/IrrlichtBaW/src/ExtImageCodec.cpp [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+ *   OTHER DEALINGS IN THE SOFTWARE.
+ ***************************************************************************/
+#include "../ext/CEGUI/IrrlichtBaW/include/ExtImageCodec.h"
+#include "../ext/CEGUI/IrrlichtBaW/include/ExtTexture.h"
+#include "CEGUI/Exceptions.h"
+#include "CEGUI/Size.h"
+#include <irrlicht.h>
+
+// Start of CEGUI namespace section
+namespace CEGUI
```
Comment (devshgraphicsprogramming) [inline ext/CEGUI/IrrlichtBaW/src/ExtImageCodec.cpp:34]: in .cpp files you can use `using namespace`

**Comment by devshgraphicsprogramming** [comment]
Don't prefix every class with `IrrlichtBaW` just put it in the `irr::ext::CEGUI` namespace

File: ext/CEGUI/IrrlichtBaW/include/ExtTextureTarget.h [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+#ifndef _CEGUISaga3DTextureTarget_h_
+#define _CEGUISaga3DTextureTarget_h_
+
+#include "../../TextureTarget.h"
+#include "CEGUI/RendererModules/Saga3D/RenderTarget.h"
+
+#if defined(_MSC_VER)
+#   pragma warning(push)
+#   pragma warning(disable : 4250)
+#endif
```
Comment (devshgraphicsprogramming) [inline ext/CEGUI/IrrlichtBaW/include/ExtTextureTarget.h:36]: don't use warning disable unless you know what they are for

**Comment by AnastaZIuk** [comment]
Fuck this shit, we have to change the tactic

**Comment by manhnt9** [comment]
Hey, be polite on code review. It's your company's image on the Internet that anybody could see!

**Comment by AnastaZIuk** [comment]
Forgive me, I have done it all wrong and got angry

**Comment by devshgraphicsprogramming** [comment]
> Hey, be polite on code review. It's your company's image on the Internet that anybody could see!

Results, meritocracy... not rhetoric ;)

As long as he gets the job done on time and on budget he can use whatever language he pleases.

Also @AnastaZIuk was referring to his own code/attempts, not yours @manhnt9  ;)

**Comment by manhnt9** [comment]
I'm not referring to anyone's work, really. Just saw 2 f words that I thought which wasn't necessary.
One from you and one from him.

**Comment by manhnt9** [comment]
> what the f are you doing deleting example 13?

 here.

**Comment by manhnt9** [comment]
I'm just minding for open-source culture on the Internet. You can treat however you like in your company.

**Comment by devshgraphicsprogramming** [comment]
> I'm just minding for open-source culture on the Internet. 

Yeah, its getting pretty annoying with all the "woke-ness"

**Comment by manhnt9** [comment]
OK, if those suggestions seems not useful to you then I have my notes here.

**Pull request closed without merging**